WIP: dex explanation (#59)
dfinity/sns-kongswap-adaptor: DEX.md

# What is a DEX?

A Decentralized EXchange (DEX) is a smart contract that enables users to trade tokens without relying on a centralized intermediary. Instead, trades are executed directly on-chain through liquidity pools.

> # Liquidity Pool:

Reviewer: ```suggestion
# Liquidity Pool (LP)
```
As we use the abbreviation later, we should introduce it once

Author: I agree that it can be quite confusing, but LP is used for Liquidity Providers. So, I think we can keep the "Liquidity Pool" non-abreviated.

Reviewer: ah if we are consistent with this throughout the text, I am ok with that!

Each DEX consists of one or more liquidity pools. A pool typically holds two tokens (say, $T_0$ and $T_1$). 

> In an ideal world, there would be a pool for every possible token pair.

Reviewer: No strong opinion, but we could also omit this line I think (except if you had something particular in mind why this is important - in which case maybe we can make this clearer?)

Role of the users interacting with a liquidity pool can be:

1. Liquidity Providers (LPs): Users who deposit tokens into pools are called liquidity providers.
2. Traders: Users who swap one token for another. The exchange rate is determined by the ratio of tokens in the pool.
3. Platform providers: Owners of the DEX.

When a trade occurs, the trader pays a fee. This fee usually consists of:
1. A platform fee that goes to the DEX protocol, and
2. An LP fee that is distributed to liquidity providers, proportional to the amount of liquidity they contributed.

Some platforms allow LPs to claim and reinvest their fees, effectively compounding their rewards.

# Market Ratio:

> A liquidity pool maintains a market ratio (MR) between its two tokens:

Reviewer: what does this mean intuitively? something like the number of one tokens must at least be half as much and at most be twice as much as the number of the other token? 
This is also basically the "price"? 

Just thought adding such an intuition might help to then parse the formula more easily

$MR = \frac{reserve_0}{reserve_1}$

> When an LP deposits $amount_0$ and $amount_1$ to the pool (deposit ratio $DR = \frac{amount_0}{amount_1}$), pool takes from $amount_0$ and $amount_1$ according to the market ratio:

Reviewer: ```suggestion
When an LP deposits $amount_0$ and $amount_1$ to the pool (deposit ratio $DR = \frac{amount_0}{amount_1}$), the pool takes from $amount_0$ and $amount_1$ according to the market ratio:
```

1. if $DR == MR$ => pool takes $amount_0$ and $amount_1$
2. if $DR >  MR$ => pool takes $amount_1 \times MR$ from $amount_0$ and $amount_0$
3. if $DR <  MR$ => pool takes $amount_0$ and $amount_0 \times MR$ from $amount_1$

This ensures that liquidity is always added proportional to the current reserves' balances.
If $DR \ne MR$, the excess amount is retunred to the LP.

# Trading:

> When trading, trader sends (without loss of generality we can think of swap $T_0$ for $T_1$) $amount_0$ of $T_0$.

Reviewer: ```suggestion
Assume a trader wants to swap $T_0$ for $T_1$. To trade, they sends $amount_0$ of $T_0$.
```

1. Liquidity Providers (LPs): Users who deposit tokens into pools are called liquidity providers.
2. Traders: Users who swap one token for another. The exchange rate is determined by the ratio of tokens in the pool.
3. Platform providers: Owners of the DEX.

When a trade occurs, the trader pays a fee. This fee usually consists of:
1. A platform fee that goes to the DEX protocol, and
2. An LP fee that is distributed to liquidity providers, proportional to the amount of liquidity they contributed.

Some platforms allow LPs to claim and reinvest their fees, effectively compounding their rewards.

# Market Ratio

A liquidity pool maintains a market ratio (MR) between its two tokens:

$MR = \frac{reserve_0}{reserve_1}$

Equivalently, it is the price of $T_1$ quoted in $T_0$. Intuitively, the higher the relative abundance of $T_0$ compared to $T_1$, the cheaper $T_1$ becomes in terms of $T_0$.

When an LP deposits $amount_0$ and $amount_1$ to the pool (deposit ratio $DR = \frac{amount_0}{amount_1}$), pool takes from $amount_0$ and $amount_1$ according to the market ratio:

1. if $DR == MR$ => pool takes $amount_0$ and $amount_1$

> 2. if $DR >  MR$ => pool takes $amount_1 \times MR$ from $amount_0$ and $amount_0$

Reviewer: what does the last part "and amount_0" mean? how much does it take from the other token?

3. if $DR <  MR$ => pool takes $amount_0$ and $amount_0 \times MR$ from $amount_1$

This ensures that liquidity is always added proportional to the current reserves' balances.
If $DR \ne MR$, the excess amount is returned to the LP.

# Trading

Assume a trader wants to swap $T_0$ for $T_1$. To trade, they send $amount_0$ of $T_0$.

> In the uniform liquidity model (used in Uniswap V2), liquidity is defined as:

Author: Used the term "popularized"..

$L^2 = reserve_0 * reserve_1$

This formula is known as *Constant Product Invariant*. When trading, the liquidity remains unchanged:

$reserve_0 * reserve_1 = (reserve_0 + amount_0) * (reserve_1 - amount_1)$

$amount_1$ is the amount of $T_1$ that is going to be removed from the pool. Before sending
it to the user, some fees are deducted: mainly a platform fee that goes to the owners of the
DEX and an LP-fee that is going to be accumulated for the liquidity providers proportional to the 
amount of liquidity they have added to the system.

# Withdrawing Liquidity

Liquidity providers can withdraw their tokens at any time. However, due to trades happening in between, they rarely receive the same token amounts they initially deposited. Instead, they withdraw, depending on the net direction of trades:

1. More of one token,
2. Less of the other, and
3. The fees they have accrued during trading activity.

Some DEXes provide the LPs with the functionality to withdraw their accrued fee and, if desired, reinvest it to
the protocol to compound their rewards.

# Pitfalls and Fallacies

> In this section, we will highlight common pitfalls and misconceptions that can be confusing for users.

Reviewer: ```suggestion
In this section, we will highlight common pitfalls and misconceptions.
```

## Pool Price $\ne$ Actual Price

> The pool exchange price (which is called *Local Price*) is solely defined by the ratio of assets in the pool. As users can trade their tokens, the ratio of the tokens diverges from the actual price.

Author: added.

## Impermanent Loss

> When LPs deposit assets to a pool, the exchange price can change according to the point $MR$. As mentioned above, the local price can be different from the actual price. Hence, if one asset, e.g., $T_0$, is significantly more appreciated than the other one. In this case, liquidity providers end up selling their assets at an undervalued price relative to the broader market. 

Reviewer: ```suggestion
When LPs deposit assets to a pool, the exchange price can change according to the point $MR$. As mentioned above, the local price can be different from the actual price. Hence, one asset, e.g., $T_0$, is significantly more appreciated than the other one. In this case, liquidity providers end up selling their assets at an undervalued price relative to the broader market. 
```
(something just seemed to be a bit off - please check if this is the right meaning)

## Slippage

> Traders don’t swap at a single fixed price, but along the curve defined by the constant product rule. Larger trades move the reserves more, so the average execution price is worse than the quoted start price.

Author: It is defined above. Maybe I should link it that section.

## Front-running

When used as an extension to SNS, each initialization, deposit, or withdraw proposal goes through a public voting process. It means that an attacker can guess the outcome of a voting when in its last stage. As explained before, they can then front-run the execution of the proposal by submitting a trade transaction and move the market ratio away from the ratio expected in the proposal.

## Liquidity Provider Risks

Although LPs are entitled to receive LP-fees, LP fees do not always offset impermanent loss. As a result, LPs may end up with fewer assets in value than if they had simply held their tokens.

## Transaction Fees

> Every deposit, withdrawal, or trade involves blockchain transaction fees. These costs reduce the effective amount received.

Reviewer: While this is true, they are very minimal on ICP usually. 
On ICP I would not call this "blockchain transaction fee" but rather "ledger transaction fee"

This is called the *Constant Product Invariant*. The key idea is that, when trades happen, the product of reserves remains unchanged:

$reserve_0 \times reserve_1 = (reserve_0 + amount_0) \times (reserve_1 - amount_1)$

While this is the most common AMM model, other variants exist (e.g., constant sum for stablecoins, hybrid curves like Curve’s StableSwap).
The adaptor is, at this stage, implemented exclusively for Kongswap. Since Kongswap (as of 5th September 2025) employs the constant product AMM, our implementation and tests are aligned with that model.

$amount_1$ is the amount of $T_1$ that is going to be removed from the pool. Before sending
it to the user, some fees are deducted: a platform fee (paid to the DEX) and an LP fee (accrued to liquidity providers, proportional to their contribution).

# Withdrawing Liquidity

LPs can withdraw their tokens at any time. However, due to trades happening in between, they rarely receive back the same token amounts they initially deposited. Instead, they withdraw, depending on the net direction of trades:

1. More of one token,
2. Less of the other, and
3. The fees they have accrued during trading activity.

> Some DEXes provide the LPs with the functionality to withdraw their accrued fee and, if desired, reinvest it to
> the protocol to compound their rewards.

Reviewer: ```suggestion
Some DEXes provide the LPs with the functionality to withdraw their accrued fee and, if desired, reinvest it to the protocol to compound their rewards.
```

# Pitfalls and Fallacies

> In this section, we will highlight common pitfalls and misconceptions.

Reviewer: ```suggestion
In this section, we will highlight some important pitfalls and facts.
```

(If we write misconsepctions, it is no clear if what we state below is now the truth or a misconseption)

## Pool Price $\ne$ Actual Price

The pool exchange price (which is called *Local Price*) is solely defined by the ratio of assets in the pool. As users can trade their tokens, the ratio of the tokens diverges from the actual price, which is determined by the broader market of Centralized Exchanges, other DEXes, OTC desks, and oracles.

## Impermanent Loss

> When LPs deposit assets to a pool, the exchange price can change according to the $MR$. As mentioned in [Pool Price $\ne$ Actual Price](#pool-price-actual-price), the local price can be different from the actual price. Hence, if one asset, e.g., $T_0$, is significantly more appreciated than the other one. In this case, liquidity providers end up selling their assets below the broader market valubelow the broader market value.

Reviewer: ```suggestion
When LPs deposit assets to a pool, the exchange price can change according to the $MR$. As mentioned in [Pool Price $\ne$ Actual Price](#pool-price-actual-price), the local price can be different from the actual price. Hence, if one asset, e.g., $T_0$, is significantly more appreciated than the other one, liquidity providers end up selling their assets below the broader market value.
```
